#!/usr/bin/env Rscript

############################## ARGUMENTS SECTION #############################
## Collect arguments
args <- commandArgs(TRUE)

## Default setting when no all arguments passed or help needed
if("--help" %in% args | "help" %in% args | (length(args) == 0) | (length(args) == 1) ) {
  cat("
      The helper R Script rsem2hbadeals.R
      Mandatory arguments:
          --rsem_folder=path        - The path to the folder with the RSEM output files. The expected filenames from RSEM are:
                                      '<sample_id>.isoforms.results', '<sample_id>.genes.results'
                                      The 'transcript_id', 'gene_id' and 'expected_count' columns MUST be present in the files generated by RSEM.
                                      '.' is accepted for denoting that the files are in the current working directory.

          --metadata=path           - A comma seperated file with metadata about the samples in the RSEM files.
                                      Rownames must be the samples. Column names must be metadata about the samples. 
                                      Two columns MUST be present in the file. One with the SRR id and one that denotes the control/case status.
                                      The default column names are for the two mandatory columsn are: 
                                      1. 'sample_id': accepted values are valid SRR ids eg. SRR10503939
                                      2. 'status': accepted values are one of {'control', 'case'}
                                      Optionally, alternative column names can be defined by the user see --sample_colname, --status_colname below.
                                      NOTE: The SRR ids must include the samples that correspond to the RSEM files provided via --rsem_folder.

          --rsem_file_suffix=value  - A string in single quotes, one of the following:
                                      'isoform.results' or 'genes.results'.
                                      
          --output=path             - A filename to save the hbadeals::hbadeals() output.
          
         --help                     - you are reading it
         
      Optional arguments:
          --sample_colname=chr      - Column name in the metadata file that contains the SRR id, default: 'sample_id'
          --status_colname=chr      - Column name in the metadata file with the control,case status information, default: 'status'
          --n_cores=int             - Number of cores for hbadeals::hbadeals(), default: 2
          --isoform_level=boolean   - TRUE or FALSE. Check ?hbadeals::hbadeals, default: TRUE
          --mcmc_warmup=int         - Number of warmup iterations for MCMC. Check ?hbadeals::hbadeals for recommended value, default: 20
          --mcmc_iter=int           - Number of iterations for MCMC. Check ?hbadeals::hbadeals for recommended value, default: 100

          --zeroes_threshold=float  - Fraction expressed as float, default: 0.9
                                      Example: 0.9 for allowing 10% of observations of the minority class to have zero counts
                                      in a transcript to not discard it from the countsData table.
          
     Usage:
      
          The typical command for running the script is as follows:
    
          ./rsem2hbadeals.R --rsem_folder='.' --metadata='./metadata.csv' --rsem_file_suffix='.isoforms.results' --output='hbadeals_results.csv'
      
      
      WARNING : here put all the things the user has to know
      \n")
  
  
  q(save="no")
}

## Parse arguments (we expect the form --arg=value)
parseArgs    <- function(x) strsplit(sub("^--", "", x), "=")

argsL        <- as.list(as.character(as.data.frame(do.call("rbind", parseArgs(args)))$V2))
names(argsL) <- as.data.frame(do.call("rbind", parseArgs(args)))$V1
args         <- argsL
rm(argsL)

## Give some value to optional arguments if not provided
if(is.null(args$sample_colname))  {args$sample_colname   = "sample_id"} else {args$sample_colname=as.character(args$sample_colname)}
if(is.null(args$status_colname))  {args$status_colname   = "status"}    else {args$status_colname=as.character(args$status_colname)}
if(is.null(args$isoform_level))   {args$isoform_level    = TRUE}        else {args$isoform_level=as.logical(args$isoform_level)}
if(is.null(args$n_cores))         {args$n_cores          = 2}           else {args$n_cores=as.numeric(args$n_cores)}
if(is.null(args$mcmc_warmup))     {args$mcmc_warmup      = 20}          else {args$mcmc_warmup=as.numeric(args$mcmc_warmup)}
if(is.null(args$mcmc_iter))       {args$mcmc_iter        = 100}         else {args$mcmc_iter=as.numeric(args$mcmc_iter)}
if(is.null(args$zeroes_threshold)){args$zeroes_threshold = 0.9}         else {args$zeroes_threshold=as.numeric(args$zeroes_threshold)}

cat("\n")
cat("ARGUMENTS SUMMARY")
cat("\n")
cat("rsem_folder      : ", args$rsem_folder     ,"\n",sep="")
cat("metadata         : ", args$metadata        ,"\n",sep="")
cat("sample_colname   : ", args$sample_colname  ,"\n",sep="")
cat("status_colname   : ", args$status_colname  ,"\n",sep="")
cat("rsem_file_suffix : ", args$rsem_file_suffix,"\n",sep="")
cat("output           : ", args$output          ,"\n",sep="")
cat("n_cores          : ", args$n_cores         ,"\n",sep="")
cat("isoform_level    : ", args$isoform_level   ,"\n",sep="")
cat("mcmc_warmup      : ", args$mcmc_warmup     ,"\n",sep="")
cat("mcmc_iter        : ", args$mcmc_iter       ,"\n",sep="")
cat("zeroes_threshold : ", args$zeroes_threshold,"\n",sep="")

# Facilitates testing
rsem_folder        <- args$rsem_folder
metadata           <- args$metadata
sample_colname     <- args$sample_colname
status_colname     <- args$status_colname
rsem_file_suffix   <- args$rsem_file_suffix
output             <- args$output
n_cores            <- args$n_cores
isoform_level      <- args$isoform_level
mcmc_warmup        <- args$mcmc_warmup
mcmc_iter          <- args$mcmc_iter
zeroes_threshold   <- args$zeroes_threshold

############################## LIBRARIES SECTION #############################

suppressWarnings(suppressMessages(library(limma)))
suppressWarnings(suppressMessages(library(coda)))
suppressWarnings(suppressMessages(library(rstan)))
suppressWarnings(suppressMessages(library(codetools)))
suppressWarnings(suppressMessages(library(hbadeals)))

# ############################### SCRIPT SECTION ###############################

# Read metadata and select required columns
accepted_status_labels_chr <- c("control", "case")

metaData  <- read.table(metadata, header = TRUE, sep = ",")
metaData  <- metaData[, c(sample_colname, status_colname)]

# Validate metadata columns for required specs
if (!(length(unique(metaData[[status_colname]]))) == 2){
    stop("The status column in the metadata table has more than 2 levels (eg. case, control, something_else).\nType ./rsem2hbadeals.R --help to check the metadata specifications.")
}
if(! (all(unique(metaData[[status_colname]]) %in% accepted_status_labels_chr)) ) {
    stop("The status column in the metadata table has values other {'control','case'}, which are the only ones accepted.\nType ./rsem2hbadeals.R --help to check the metadata specifications.")
}

# Convert to control,case to 1,2 as requested in hbadeals::hbadeals()
metaData[[status_colname]] <- ifelse( metaData[[status_colname]] == "control", 1, 2)
colnames(metaData)  <- c("sample_id", "status")

# Get paths of RSEM files
isoform_individual_files <- list.files(rsem_folder,
                                       full.names = TRUE,
                                       pattern    = rsem_file_suffix)

# Keep 'gene_id' and 'transcript_id' columns only (scaffold to build collective countsData csv from all samples)
countsData  <- read.table(isoform_individual_files[1], header=TRUE)
countsData  <- countsData[, c("transcript_id", "gene_id")]

# Iteratively append each individual sample RSEM 'expected_count' column to create cohort table
for (input.file in isoform_individual_files  ) {
	next.file  <- read.table(input.file, header=TRUE)
	sample_id  <- gsub(rsem_file_suffix, "", basename(input.file))
	
	# Heuristic to name a column after the sample id
	next.file[[sample_id ]] <- next.file$expected_count
	toKeep     <- c("transcript_id", sample_id)
	countsData <- merge(countsData, next.file[, toKeep], by = "transcript_id")
}

# Calculating sample size
sample_size <-  dim(countsData)[2] - 2
# message(paste0('Sample size = ', sample_size))

# Temporarily dettach metadata columns "transcript_id","gene_id" from numerical values
all      <- colnames(countsData)
toRemove <- c("transcript_id","gene_id")
toKeep   <- all [!all  %in% toRemove]

# Apply exclusion criterion 1:
# Discard transcripts with zero counts across samples (zeroes tolerance threshold on minority class is set to zeroes_threshold)
message("\nApplying exclusion criterion 1: discard transcripts with zero counts across samples (zeroes tolerance threshold on minority class is set to ", zeroes_threshold, ")")

N_controls       <- length(metaData[[status_colname]][metaData[[status_colname]] == 1])
N_cases          <- length(metaData[[status_colname]][metaData[[status_colname]] == 2])
N_minority_class <- min(N_controls, N_cases)

message('N controls       : ', N_controls       , appendLF = FALSE)
message('N cases          : ', N_cases          , appendLF = FALSE)
message('N minority class : ', N_minority_class , appendLF = FALSE)

control_ids <- metaData[[sample_colname]][metaData[[status_colname]] == 1]
cases_ids   <- metaData[[sample_colname]][metaData[[status_colname]] == 2]

message('dim(countsData)[1] before transcript exclusion: ', dim(countsData)[1])

message("\nApplying exclusion criterion 1: discard transcripts with zero counts across samples (zeroes tolerance threshold on minority class is set to ", zeroes_threshold, ")")
countsData <- countsData[rowSums(countsData[, colnames(countsData) == control_ids ] > 0) > N_minority_class * zeroes_threshold, ]
message('dim(countsData)[1] after applying transcript exclusion based on control: ', dim(countsData)[1])

countsData <- countsData[rowSums(countsData[, colnames(countsData) == cases_ids   ] > 0) > N_minority_class * zeroes_threshold, ]
message('dim(countsData)[1] after applying transcript exclusion based on cases: ', dim(countsData)[1])


message("Done!")

# Apply exclusion criterion 2: Filter out transcripts with only one isoform
message("\nApplying exclusion criterion 2: ( keep > 1 isoform transcripts) ..")
num.iso=unlist(lapply(countsData$gene_id, function(x){sum(countsData$gene_id %in% x)}))
message('dim(countsData)[1] before gene exclusion based on number of isoforms: ', dim(countsData)[1])
countsData <- countsData[num.iso > 1, ]
message('dim(countsData)[1] before gene exclusion based on number of isoforms ( > 1): ', dim(countsData)[1])

# Reorder countsData columns by metaData order !DANGEROUS TO RELY ON INDEXES!
# message("\nOrdering 'countsData' columns according to 'metaData$sample_id' order ..")
toKeepInOrder <- c( "gene_id", "transcript_id", as.vector(metaData$sample_id) )
countsData    <- countsData[, toKeepInOrder]
# message("Done!")
write.table(countsData,
            file      = paste0("countsData_", output, ".csv"),
            sep       = ',',
            quote     = F,
            col.names = T,
            row.names = F)

# Run HBA-DEALS
# message('\nRunning hbadeals::hbadeals(). NOTE: This step can take a while ..')
res <-hbadeals::hbadeals(countsData     = countsData,
                          labels        = as.factor(metaData$status),
                          n.cores       = n_cores,
                          isoform.level = isoform_level,
                          mcmc.iter     = mcmc_iter,
                          mcmc.warmup   = mcmc_warmup)
# message('Successfully completed hbadeals::hbadeals() task!')

# Write results in file
# message('\nWriting  hbadeals::hbadeals() results to a file..')
write.table(res,
            file = paste0(output, ".csv"),
            sep       = ',',
            quote     = F,
            col.names = T,
            row.names = F)
# message('Done!')
# message('\nGoodbye!')
